### PR TITLE
FRMW-681: Add 10-year range to rxDatePicker

### PR DESCRIPTION
**Pull request by 100stacks**
JIRA: https://jira.rax.io/browse/FRMW-681

### LGTMs
- [x] Dev LGTM
- [x] Dev LGTM
- [x] Design LGTM
- [x] QE LGTM

---
Refactor `rxDatePicker` to provide a 10-year date range before and after current date.

### CODE REVIEW

**Comment by 100stacks** [comment]
## `rxDatePicker` Demo Page

### Before with Static `Month` and `Year`
rxdatepicker - encoreui 2016-04-07 11-41-38

### After with `Month` and `Year` Dropdowns
rxdatepicker - encoreui 2016-04-07 11-46-38

screen shot 2016-04-07 at 11 46 51 am

screen shot 2016-04-07 at 11 47 13 am

**Comment by CITguy** [comment]
Please note, that the year range is calculated from the _selected_ date. If there is no selection, it will be calculated from the _current_ date.

File: src/molecules/rxDatePicker/docs/rxDatePicker.midway.js [hunk 1]
```diff
@@ -41,10 +46,8 @@ describe('rxDatePicker', function () {
         });
 
         it('should display the current month and year by default', function () {
-            picker.monthAndYear.then(function (date) {
-                expect(date.getYear()).to.equal(today.getYear());
-                expect(date.getMonth()).to.equal(today.getMonth());
-            });
+            expect(picker.month).to.eventually.equal(month);
```
Comment (parlarjb) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: Yay for moving `expect` out of the `.then()`! 💯

File: src/molecules/rxDatePicker/scripts/rxDatePicker.js [hunk 2]
```diff
@@ -34,6 +34,7 @@ angular.module('encore.ui.molecules')
         link: function (scope, element, attrs, ngModelCtrl) {
             var today = moment(new Date());
             var isoFormat = 'YYYY-MM-DD';
+            const YEAR_RANGE = 10;
```
Comment (parlarjb) [inline src/molecules/rxDatePicker/scripts/rxDatePicker.js]: Could you move this to before the `return` on line 27? Since it doesn't change, we don't need to redefine it every time the linking function is called.

Comment (parlarjb) [inline src/molecules/rxDatePicker/scripts/rxDatePicker.js]: And the `isoFormat` could do with the same treatment

File: src/molecules/rxDatePicker/scripts/rxDatePicker.js [hunk 3]
```diff
@@ -168,6 +200,26 @@ angular.module('encore.ui.molecules')
...
                 return days;
             }//buildCalendarDays
+
+            /**
+             * @param {Moment} midpoint
+             * @return {Array<ISO 8601 Year> }
+             * description
+             * Generate an array of ISO 8601 Year (format "YYYY") years.
+             */
+            function generateCalendarYears (midpoint) {
```
Comment (parlarjb) [inline src/molecules/rxDatePicker/scripts/rxDatePicker.js]: We could also move this function and `buildCalendarDays` outside of the `return` on 27. As it stands now, we recreate these functions every time the linking function is called.

File: src/molecules/rxDatePicker/docs/rxDatePicker.midway.js [hunk 4]
```diff
@@ -9,8 +9,13 @@ describe('rxDatePicker', function () {
 
     describe('simple example', function () {
         var today = new Date();
-        var lastMonth = new Date(moment(today).subtract(1, 'month'));
+        var month = today.toString().split(' ')[1];
+        var year = today.toString().split(' ')[3];
+        var lastMonth = new Date(moment(today).subtract(1, 'month').startOf('month'));
```
Comment (CITguy) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: This line seems confusing. Can we convert to using moment objects only?

File: src/molecules/rxDatePicker/scripts/rxDatePicker.page.js [hunk 5]
```diff
@@ -317,6 +294,26 @@ var rxDatePicker = {
...
+        get: function () {
+            return exports.rxSelect.initialize(this.rootElement.element(by.model('currentMonth'))).selectedOption.text;
+        },
+
+        set: function (value) {
+            exports.rxSelect.initialize(this.rootElement.element(by.model('currentMonth'))).select(value);
+        }
+    },
+
+    year: {
```
Comment (CITguy) [inline src/molecules/rxDatePicker/scripts/rxDatePicker.page.js]: It might be worth noting that the year is calculated from the selected date, if present. Otherwise, it is calculated from today's date.

File: src/molecules/rxDatePicker/scripts/rxDatePicker.js (hunk 3 above)
Comment (CITguy) [inline src/molecules/rxDatePicker/scripts/rxDatePicker.js]: These functions might even be candidates to be placed in a `rxDatePickerUtil` service for testing purposes.  This may be out of scope for the current PR, though.

File: src/molecules/rxDatePicker/docs/rxDatePicker.midway.js (hunk 1 above)
Comment (100stacks) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: @parlarjb @CITguy should I remove all midways `expect` clauses out of `then` statements?

Comment (parlarjb) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: When possible, yes. But I know that Andrew recently had trouble with that. It's easier in unit tests where you can do an explicit `$digest()` to ensure that all promises have been resolved, but apparently it's a lot tougher in midways.

The reason you'd want to move them is I've seen it happen _multiple_ times where an `expect()` is inside a `.then()`, but the promise never resolves, and thus the `expect()` never gets executed. This appears to be a successful test, when in fact the test never really ran.

Comment (100stacks) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: Okay, so promises may throw false positives.  :confused:

I paired up with @Droogans on the refactoring the page objects and midways, so any other feedback is always appreciated! :+1:

Comment (Droogans) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: It's not quite that simple.

```js
expect(picker.month).to.eventually.equal(month);
                        ^^^^^^^^^^
```

That's syntactic sugar for this.

```js
picker.month.then(function (m) {
    expect(m).to.equal(month);
});
```

File: src/molecules/rxDatePicker/scripts/rxDatePicker.js (hunk 2 above)
Comment (Droogans) [inline src/molecules/rxDatePicker/scripts/rxDatePicker.js]: Also, you can't use `const` like that yet.

We're adhering to strict node v0.12.x syntax until protractor 3.x is merged in, which at that point we'll start using node 4+ syntax like this in the page object code.

File: src/molecules/rxDatePicker/docs/rxDatePicker.midway.js (hunk 4 above)
Comment (Droogans) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: Moment objects don't support formatting for short months, do they? I'll have to check.

File: src/molecules/rxDatePicker/scripts/rxDatePicker.js (hunk 2 above)
Comment (CITguy) [inline src/molecules/rxDatePicker/scripts/rxDatePicker.js]: @Droogans, use `var` instead?

File: src/molecules/rxDatePicker/docs/rxDatePicker.midway.js (hunk 4 above)
Comment (CITguy) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: http://momentjs.com/docs/#/displaying/format/

File: src/molecules/rxDatePicker/scripts/rxDatePicker.js (hunk 2 above)
Comment (Droogans) [inline src/molecules/rxDatePicker/scripts/rxDatePicker.js]: Yes please.

File: src/molecules/rxDatePicker/scripts/rxDatePicker.page.js (hunk 5 above)
Comment (Droogans) [inline src/molecules/rxDatePicker/scripts/rxDatePicker.page.js]: This is straight up just not documented.

My bad, I paired with @100stacks on this and we were more focused on getting end to end code refactored and tests passing.

**Comment by glynnis** [comment]
Looking closely at your work locally and comparing with the mock-ups from JIRA, there's one small tweak to make, which is putting a few pixels of spacing between the month and year dropdowns:

### JIRA mockup
screen shot 2016-04-07 at 2 23 04 pm

### Work from this PR
screen shot 2016-04-07 at 2 23 22 pm

Another thing of note is the gray circle around the prior-selected date doesn't seem to update properly. On page load, April 7 is the date selected. Even after making multiple updates to the date, the gray circle still appears around April 7. 
#### Open dropdown, select April 13 and click away from date picker (gray circle still visible as expected)
screen shot 2016-04-07 at 2 23 22 pm

#### Select a date in May, click away from date picker
screen shot 2016-04-07 at 2 29 38 pm

#### Open date picker and page back to April, gray circle visible despite subsequent selections
screen shot 2016-04-07 at 2 29 34 pm

I would expect the gray circle to update to show the most recent selection before change, rather than remember the first selection on page load.

The mock-ups in JIRA don't show any gray circles at all, but I'm guessing this functionality was already a part of rxDatePicker before this PR, correct?

**Comment by 100stacks** [comment]
@glynnis, I think the gray circle represents the current day of year local so in this case April 7, 2016.

File: src/molecules/rxDatePicker/rxDatePicker.less [hunk 6]
```diff
@@ -137,7 +147,8 @@ rx-date-picker {
           }
         }
 
-        &.currentMonth {
+        &.yearWrapper,
+        &.monthWrapper {
           flex: 1;
```
Comment (glynnis) [inline src/molecules/rxDatePicker/rxDatePicker.less]: This is where you can add a margin. `margin-right: 5px;` looks about right to me.

Comment (glynnis) [inline src/molecules/rxDatePicker/rxDatePicker.less]: Might be worth putting it only on `.monthWrapper`, though, so you're not adding margin to the year too.

**Comment by 100stacks** [comment]
okpv_tinygif

**Comment by 100stacks** [comment]
@glynnis, followed your `margin-right: 5px;` on `.monthWrapper` recommendation in https://github.com/rackerlabs/encore-ui/pull/1619#discussion_r58940292.

screen shot 2016-04-07 at 4 30 13 pm

**Comment by 100stacks** [comment]
@parlarjb @CITguy, feedback incorporated.

**Comment by 100stacks** [comment]
Updated `rxDatePicker` screenshots and demo gif in https://github.com/rackerlabs/encore-ui/pull/1619#issuecomment-206990774.

**Comment by 100stacks** [comment]
@glynnis brings up a good point in https://github.com/rackerlabs/encore-ui/pull/1619#issuecomment-207065397 concerning the gray circle around current day, do we need to annotate its purpose for users?

File: src/molecules/rxDatePicker/docs/rxDatePicker.html [hunk 7]
```diff
@@ -25,7 +25,7 @@
   <p>
     Sometimes, a form may need to prepopulate a value for rxDatePicker. The
     example below shows how the element behaves when its model is defaulted
-    to <code>2015-11-17</code>.
+    to today's date.
```
Comment (glynnis) [inline src/molecules/rxDatePicker/docs/rxDatePicker.html]: @100stacks If you want to make a note about the gray circle, this might be a good place. Something like, `When a different date is selected, a gray circle around the current date provides additional context to users as they find their selection in the date picker.`

**Comment by glynnis** [comment]
LGTM

File: src/molecules/rxDatePicker/rxDatePicker.less [hunk 8]
```diff
@@ -137,8 +147,13 @@ rx-date-picker {
           }
         }
 
-        &.currentMonth {
+        &.yearWrapper {
```
Comment (glynnis) [inline src/molecules/rxDatePicker/rxDatePicker.less]: Is there a good reason for keeping these camel case? I thought our standards called for kebab case, with a few rare exceptions.

I read over our CSS style guide (http://rackerlabs.github.io/encore-ui/#/guides/css-styleguide), but don't fully understand why we'd use camel case here, unless this falls under the "attribute directive" rules at the bottom? If so, I think we could add a few more examples to the style guide that aren't just camel case classes based on the directive names themselves.

@CITguy can you help clarify?

**Comment by CITguy** [comment]
@glynnis, good catch. Yes, we should change those classes to kebab case.

@100stacks ^^

**Comment by 100stacks** [comment]
@glynnis @CITguy, changed to kebab case: `month-wrapper` and `year-wrapper` . :+1:

File: src/molecules/rxDatePicker/docs/rxDatePicker.html [hunk 9]
```diff
@@ -25,7 +25,9 @@
   <p>
     Sometimes, a form may need to prepopulate a value for rxDatePicker. The
     example below shows how the element behaves when its model is defaulted
-    to <code>2015-11-17</code>.
+    to today's date.  When a different date is selected, a gray circle around
+    the current date provides additional context to users as they find their
+    selection in the date picker.
```
Comment (100stacks) [inline src/molecules/rxDatePicker/docs/rxDatePicker.html]: @glynnis, added your feedback for users. 

screen shot 2016-04-08 at 11 59 48 am

File: src/components/rxSelect/scripts/rxSelect.page.js [hunk 10]
```diff
@@ -256,6 +256,7 @@ var rxSelect = {
     },
 
     /**
+     * @todo: Stop returning a select option object -- I just want the text!
```
Comment (halkon) [inline src/components/rxSelect/scripts/rxSelect.page.js]: this makes me curious, I could see cases for both, can you describe why one over the other?

File: src/molecules/rxDatePicker/scripts/rxDatePicker.js [hunk 11]
```diff
@@ -24,6 +24,64 @@ angular.module('encore.ui.molecules')
...
+            days.push(appendDay.add(1, 'day'));
+        }
+
+        return days;
+    }//buildCalendarDays
+
+    /**
+     * @param {Moment} midpoint
+     * @return {Array<ISO 8601 Year> }
+     * description
```
Comment (CITguy) [inline src/molecules/rxDatePicker/scripts/rxDatePicker.js]: `description` -> `@description`

File: src/molecules/rxDatePicker/scripts/rxDatePicker.page.js (hunk 5 above)
Comment (CITguy) [inline src/molecules/rxDatePicker/scripts/rxDatePicker.page.js]: Please document `month` and `year`

File: src/components/rxSelect/scripts/rxSelect.page.js (hunk 10 above)
Comment (Droogans) [inline src/components/rxSelect/scripts/rxSelect.page.js]: Principle of least astonishment.

> I want to set a dropdown value -- I'll give you the text, you set it. :+1:

> I want to get a dropdown value -- I'll get back...[the option object](https://m.popkey.co/4340ce/KMzYj.gif)?! What? 

If you want the `value` attribute instead, roll your own getter/setter pair. For nearly every use case, the text in the dropdown is what the user wants, therefore, that's what the selenium tests want.

Comment (CITguy) [inline src/components/rxSelect/scripts/rxSelect.page.js]: To add to what @Droogans said, page object apis should accept/return values that are visible to an end user. (e.g. A User isn't going to be constructing a `Date` object to select a value in a date picker, so why should the page object expect it?) 

In this case, the User isn't going to be viewing the value through an `object`, but rather a simple string in the interface. So the page object should do the same.

Comment (Droogans) [inline src/components/rxSelect/scripts/rxSelect.page.js]: There are some exceptions. Many times what a user sees is `"N/A"` but what they really mean is something that is `null`. Other times, a user sees `$1.55`, when they really want something that you can perform arithmetic on.

Dates work fine in this example because you've chosen the appropriate string representation -- you can compare them, sort them, etc., and it'll all "just work". Using `'Jun 23, 2016'` is much worse than using `new Date('Jun 23, 2016')`, but not as good as `'2106-06-23'`.

File: src/molecules/rxDatePicker/docs/rxDatePicker.midway.js [hunk 12]
```diff
@@ -79,11 +79,8 @@ describe('rxDatePicker', function () {
 
         it('should reopen the calendar and have the month unchanged', function () {
             picker.open();
-            picker.monthAndYear.then(function (date) {
-                expect(date.getYear()).to.equal(nextMonth.getYear());
-                expect(date.getMonth()).to.equal(nextMonth.getMonth());
-                picker.previousMonth();
-            });
+            expect(today.get('year')).to.equal(nextMonth.get('year'));
```
Comment (CITguy) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: Are we testing the right thing here? This assertion is likely to fail in December, because `today` and `nextMonth` are dynamic based on today's date. We're not testing anything in the interface.

File: src/molecules/rxDatePicker/scripts/rxDatePicker.page.js [hunk 13]
```diff
@@ -341,7 +338,8 @@ var rxDatePicker = {
 
         set: function (date) {
             this.open();
-            this.monthAndYear = date;
+            this.month = date.toString().split(' ')[1];
+            this.year = date.toString().split(' ')[3];
```
Comment (CITguy) [inline src/molecules/rxDatePicker/scripts/rxDatePicker.page.js]: Can we update the `date` property to accept/return a `String` instead of a `Date`?

Comment (Droogans) [inline src/molecules/rxDatePicker/scripts/rxDatePicker.page.js]: We decided to put that in a separate PR.

File: src/molecules/rxDatePicker/scripts/rxDatePicker.page.js [hunk 14]
```diff
@@ -142,26 +102,41 @@ var rxDatePicker = {
...
+        value: function () {
+            var self = this;
+            return protractor.promise.all([this.year, this.month]).then(function (results) {
+                var currentDate = new Date(results[0] + ' ' + results[1]);
+                var firstDate = moment(new Date(currentDate)).startOf('month').format();
+                var firstDayString = firstDate.split('T')[0];
+                self.selectVisibleDate(firstDayString);
+            });
+        }
+    },
```
Comment (Droogans) [inline src/molecules/rxDatePicker/scripts/rxDatePicker.page.js]: @100stacks I don't think this method is necessary. The only reason the `selectLastDayOfCurrentMonth` exists is due to timezone issues making it difficult to do.

This method doesn't simplify or save any time. It should be removed.

File: src/molecules/rxDatePicker/docs/rxDatePicker.midway.js (hunk 12 above)
Comment (Droogans) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: @100stacks He's right.

We're using a `debug=true` flag in all demo page urls. Take advantage of this by `ng-show`'ing a date picker that always has the same date (middle of the year is probably best for previous/next month tests).

This will reduce the amount of work being done in the test code for accounting for all possible test dates that occur over time.

Comment (CITguy) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: Right now, this test is testing variables that are explicitly set in the `*.midway.js`. Nowhere in this test case is it asserting against the picker month or year seen in the browser.

Comment (Droogans) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: Right, that's because it's using today's date by default.

If we were to create a date picker that always had the same date

```js
expect(picker.month).to.eventually.equal(today.get('month'));
```

would become

```js
expect(picker.month).to.eventually.equal('Jun');
```

File: src/molecules/rxDatePicker/scripts/rxDatePicker.page.js (hunk 13 above)
Comment (100stacks) [inline src/molecules/rxDatePicker/scripts/rxDatePicker.page.js]: @Droogans, is this story still applicable:
https://jira.rax.io/browse/FRMW-715

Or should we capture @CITguy's request in another story?

> Can we update the date property to accept/return a String instead of a Date?

File: src/molecules/rxDatePicker/docs/rxDatePicker.midway.js [hunk 15]
```diff
@@ -79,11 +79,9 @@ describe('rxDatePicker', function () {
...
         it('should reopen the calendar and have the month unchanged', function () {
             picker.open();
-            picker.monthAndYear.then(function (date) {
-                expect(date.getYear()).to.equal(nextMonth.getYear());
-                expect(date.getMonth()).to.equal(nextMonth.getMonth());
-                picker.previousMonth();
-            });
+            expect(moment(picker.date).get('month')).to.equal(today.get('month'));
+            expect(moment(picker.date).get('year')).to.equal(today.get('year'));
+            picker.previousMonth();
```
Comment (100stacks) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: @CITguy @Droogans, I updated this test to use `today`, since I didn't understand why it would check `nextMonth`.  

```js
expect(moment(picker.date).get('month')).to.equal(today.get('month'));
expect(moment(picker.date).get('year')).to.equal(today.get('year'));
```

Comment (CITguy) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: @100stacks:
The `picker.date` property returns the selected date of the picker, which is not what we want to test.

In the series of tests before this, we open the picker and test that navigation between calendar months works both forward and backward. This test should check that by closing and reopening the picker, the calendar month has not changed. Since, the state of the calendar month at this point in the test suite is one month ahead, we should test against `nextMonth`.

Please update to test against `picker.month` and `picker.year`.

Comment (CITguy) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: The properties of the page object are confusing in this case. The `date`, `month`, and `year` properties correspond to two different aspects of the picker. `date` is actually the _value_ of the picker, while `month` and `year` are properties associated with calendar navigation.

Comment (Droogans) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: I don't disagree. If there's going to be a distinction, it should be present only to facilitate our testing needs. There's no need to expose the separate entities for the last confirmed date set, and the _current_ date (no set yet).

Public functionality should only be concerned with getting the last confirmed set date, and changing it, which is how this works currently.

File: src/molecules/rxDatePicker/docs/rxDatePicker.midway.js [hunk 16]
```diff
@@ -79,11 +79,9 @@ describe('rxDatePicker', function () {
...
         it('should reopen the calendar and have the month unchanged', function () {
             picker.open();
-            picker.monthAndYear.then(function (date) {
-                expect(date.getYear()).to.equal(nextMonth.getYear());
-                expect(date.getMonth()).to.equal(nextMonth.getMonth());
-                picker.previousMonth();
-            });
+            expect(picker.month).to.eventually.equal(nextMonthName);
+            expect(picker.year).to.eventually.equal(nextMonth.year().toString());
+            picker.previousMonth();
```
Comment (100stacks) [inline src/molecules/rxDatePicker/docs/rxDatePicker.midway.js]: @CITguy @Droogans, updated test to check `picker.month` and `picker.year`. 

`nextMonth.year()` returns a number http://momentjs.com/docs/#/get-set/year/

**Comment by CITguy** [comment]
LGTM

**Comment by parlarjb** [comment]
LGTM

**Comment by Droogans** [comment]
This has got to be the highest LOC/comment ratio I've seen in a while
